### PR TITLE
add compilerName option, and add new compiler JDTCompilerParallel

### PR DESCRIPTION
**Pull request by tcollignon**
This PR introduce new goal parameter : compilerName (default value is "org.apache.jasper.compiler.JDTCompiler")
This allow anyone to use another Compiler if he want.
I add a new compiler based on JDTCompiler, but less synchronized, and so with better performance when they are threads in parallel. I have do many tests with java mission control to try to improve JDT compiler.
I will write a little blog post I have time, to explain this and show my result.

### CODE REVIEW

**Comment by leonardehrenfried** [comment]
Wow, this is pretty great work!

Due to the lack of synchronization, is using the parallel compiler more risky? Is there a chance of a race condition?

Also, shouldn't the name of the class be `ParallelJDTCompiler`?

**Comment by tcollignon** [comment]
Hi,

It's more risky for now, because we don't have experience report, but in fact I have juste add a cache with a map (synchronized map), so there is still synchronization but it happens less often, and the performance are better. (I have change another API to avoid synchronization but it's a small things). I don't think there is race condition. 
Maybe in the future we can put this implementation by default in case of threads > 1, but we need time to be sure this is correct.
Ok for the name I change that

Thank you

File: src/main/java/io/leonard/maven/plugins/jspc/JspCContextAccessor.java [hunk 1]
```diff
@@ -36,14 +34,35 @@ protected ClassLoader getLoader() {
...
+
+  public void setResourcesCache(Map<String, NameEnvironmentAnswer> resourcesCache) {
+    this.resourcesCache = resourcesCache;
+  }
+
+  public String getCompilerName() {
+    return compilerName;
+  }
+
+  public void setCompilerName(String compilerName) {
```
Comment (leonardehrenfried) [inline src/main/java/io/leonard/maven/plugins/jspc/JspCContextAccessor.java]: Should this property not be called `compilerClass`?

File: src/main/java/io/leonard/maven/plugins/jspc/JspcMojo.java [hunk 2]
```diff
@@ -246,6 +229,16 @@
...
    */
   private String compilerVersion;
+  
+  /**
+   * Name of the compiler class used to compile the jsp files.
+   * If threads parameter is greater than 4, then maybe the compilerName "org.apache.jasper.compiler.ParallelJDTCompiler" will be more efficient
+   *
+   * @parameter default-value="org.apache.jasper.compiler.JDTCompiler"
+   */
+  private String compilerName;
```
Comment (leonardehrenfried) [inline src/main/java/io/leonard/maven/plugins/jspc/JspcMojo.java]: Same as above. I think the correct name is compilerClass.

**Comment by leonardehrenfried** [comment]
Ok, if we disable it by default I'm happy to merge this and wait if people report bugs. You're right: If it turns out to be stable enough we can enable it for multiple threads.

Also, I have a small comment regarding the new property name.

File: src/main/java/io/leonard/maven/plugins/jspc/JspcMojo.java (hunk 2 above)
Comment (tcollignon) [reply src/main/java/io/leonard/maven/plugins/jspc/JspcMojo.java]: Yes you're right

**Comment by leonardehrenfried** [comment]
Thanks, I will merge this in at the weekend and release a new version.

**Comment by tcollignon** [comment]
Thx its great. After that I Will create the 2.X branch to maintain compatibility, and start working with tomcat 9 on master branch on 3.X version

**Comment by leonardehrenfried** [comment]
Version 2.3.4 has been released.

**Comment by tcollignon** [comment]
Note : I write a little blog post on this : https://tcollignon.github.io/2018/02/19/How-to-speed-up-JSP-compile-with-Tomcat-and-Maven-20.html

**Pull request merged**